waku/v2/peermanager/rtt_cache.go

```diff
@@ -0,0 +1,256 @@
+	PingRTT time.Duration
+
+	bkf              backoff.BackoffStrategy
+	nextVerification time.Time
+	timesource       timesource.Timesource
+}
+
+func (pr *PeerRTT) IsStale() bool {
+	return pr.nextVerification.Before(pr.timesource.Now())
+}
+
+type RTTCache struct {
+	sync.RWMutex
+
+	host host.Host
+
+	peers     map[peer.ID]*PeerRTT
+	pingQueue PingQueue
+
+	verificationInterval time.Duration
+
+	timesource timesource.Timesource
+	logger     *zap.Logger
+}
+
+func NewRTTCache(timesource timesource.Timesource, logger *zap.Logger) *RTTCache {
+	return &RTTCache{
+		timesource: timesource,
+		logger:     logger.Named("ping-manager"),
+		peers:      make(map[peer.ID]*PeerRTT),
+	}
+}
+
+func (r *RTTCache) SetHost(h host.Host) {
+	r.host = h
+}
+
+func (r *RTTCache) Start(ctx context.Context) {
+	go r.start(ctx, 10*time.Second)
+}
+
+func (r *RTTCache) pingPeers(ctx context.Context) {
+	r.RLock()
+	var toPing peer.IDSlice
+	if len(r.pingQueue) == 0 {
+		toPing = r.host.Peerstore().Peers()
+	} else {
+		for _, p := range r.pingQueue {
+			if p.IsStale() {
+				toPing = append(toPing, p.PeerID)
+			}
+		}
+	}
+	r.RUnlock()
+	for _, p := range toPing {
+		go r.PingPeer(ctx, p)
+	}
+}
+
+func (r *RTTCache) start(ctx context.Context, pingInterval time.Duration) {
+	ticker := time.NewTicker(pingInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			r.pingPeers(ctx)
+		}
+	}
+
+}
+
+func (r *RTTCache) PingPeer(ctx context.Context, peer peer.ID) (*PeerRTT, error) {
+	if peer == r.host.ID() {
+		return nil, nil // Don't ping yourself
+	}
+
+	r.RLock()
+	peerRTT, ok := r.peers[peer]
+	r.RUnlock()
+
+	if ok && !peerRTT.IsStale() {
+		// We already have a ping recorded
+		return peerRTT, nil
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+
+	case result := <-ping.Ping(ctx, r.host, peer):
+		r.Lock()
+		defer r.Unlock()
+
+		peerRTT, exists := r.getOrDefault(peer)
+		if result.Error == nil {
+			peerRTT.PingRTT = result.RTT
+			peerRTT.nextVerification = time.Now().Add(r.verificationInterval)
+			if exists {
+				r.pingQueue.Update(peerRTT)
+			} else {
+				r.pingQueue.Push(peerRTT)
+			}
+			return peerRTT, nil
+		} else {
+			peerRTT.PingRTT = time.Hour // This is just so we don't choose this peer again since the RTT is unreasonable
+			peerRTT.nextVerification = time.Now().Add(peerRTT.bkf.Delay())
+			if exists {
+				r.pingQueue.Update(peerRTT)
+			} else {
+				r.pingQueue.Push(peerRTT)
+			}
+
+			r.logger.Debug("could not ping", logging.HostID("peer", peer), zap.Error(result.Error))
+
+			return nil, result.Error
+		}
+	}
+}
+
+func (r *RTTCache) PeersRTT(ctx context.Context, peers peer.IDSlice) []*PeerRTT {
+	wg := sync.WaitGroup{}
+	pingCh := make(chan *PeerRTT, len(peers))
+	wg.Add(len(peers))
+	for _, p := range peers {
+		go func(p peer.ID) {
+			defer wg.Done()
+			peerRTT, err := r.PingPeer(ctx, p)
```

Comment:
Wondering if we need to ping the peers during selection, since we have a periodic ping going on which should update the RTT for the peers. This may cause problems even if 1 peer is stale. 
Wouldn't it be better to be optimistic and return the peer with shortest ping for which we already have RTT recorded?

Reply:
For sure it would simplify the code

Reply:
Then let's keep it simple then and not do ping during selection :)

Reply:
I did a change in the code, which still requires doing the ping but the code is less complex now